File: backend/src/main/java/com/joying/chat/config/WebSocketConfig.java
package com.joying.chat.config;

import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import org.springframework.messaging.simp.config.ChannelRegistration;
import org.springframework.messaging.simp.config.ChannelRegistration;
import org.springframework.messaging.simp.config.MessageBrokerRegistry;
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker;
import org.springframework.web.socket.config.annotation.StompEndpointRegistry;
import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer;
import org.springframework.web.socket.config.annotation.WebSocketTransportRegistration;

import lombok.RequiredArgsConstructor;

/**
 * 웹소켓 STOMP 설정.
 *
 * <p>브라우저가 웹소켓을 못 쓸 때를 대비해 SockJS 폴백을 켠다.
 */
@Configuration
@EnableWebSocketMessageBroker
@RequiredArgsConstructor
public class WebSocketConfig implements WebSocketMessageBrokerConfigurer {

	@Value("${cors.allowed-origins}")
	private List<String> allowedOrigins;

	private final WebSocketAuthInterceptor webSocketAuthInterceptor;

	/**
	 * 브로커 설정.
	 *
	 * <p>주석에는 오래도록 SimpleBroker를 제거했다고 적혀 있었다. 그런데
	 * {@code enableSimpleBroker}도 {@code enableStompBrokerRelay}도 부르지 않으면
	 * Spring이 SimpleBroker를 자동으로 켠다. 브로커는 그대로 살아 있었고 바뀐 것은
	 * 목적지 접두사 제한이 사라진 것뿐이었다.
	 *
	 * <p>서버를 넘어야 하는 전달은 {@code ChatBroadcaster}가 Redis로 내보내고, 각
	 * 서버가 받아서 자기에게 붙은 세션에만 내보낸다. 여기 브로커는 그 마지막
	 * 한 걸음을 맡는다.
	 */
	@Override
	public void configureMessageBroker(MessageBrokerRegistry registry) {
		// 클라이언트가 보낼 때 붙이는 접두사. 예: SEND /app/chat/123/send
		registry.setApplicationDestinationPrefixes("/app");

		// 특정 사용자에게만 보낼 때 쓴다. 예: /user/{id}/queue/messages
		registry.setUserDestinationPrefix("/user");

		// 한 사람에게 나가는 메시지를 보낸 순서대로 내보낸다.
		//
		// 켜지 않으면 송신 채널이 스레드 풀이라 같은 사람에게 갈 두 건이 서로
		// 앞지른다. 부하를 넣고 재 보니 200건 중 193번 번호가 뒤로 갔다.
		//
		// 값은 처리량이다. 앞엣것이 끝나야 뒤엣것을 보내므로 한 사람에게 몰릴 때
		// 그 사람의 줄만 길어진다. 다른 사람은 그대로 병렬이다.
		registry.setPreservePublishOrder(true);
	}

	/**
	 * 들어오는 프레임을 연결 단위로 한 줄로 세운다.
	 *
	 * <p>기본 채널은 스레드 풀이라 한 연결에서 연달아 보낸 프레임이 서로 앞지른다.
	 * 그러면 나중에 보낸 것이 먼저 번호를 받아, 보낸 사람 화면에서도 자기 말의
	 * 앞뒤가 바뀐다. 부하를 넣기 전에는 보이지 않았다.
	 *
	 * <p>연결로 묶으므로 사람이 많아져도 서로 기다리지 않는다. 한 연결 안에서만
	 * 순서를 지킨다.
	 */
	@Override
	public void registerStompEndpoints(StompEndpointRegistry registry) {
		// 주소를 둘 받는다.
		//
		// 앞단은 방 번호로 노드를 고른다. 그 값을 쿠키로 알렸는데 쿠키는 브라우저에
		// 하나뿐이라, 탭을 둘 열어 서로 다른 방을 보면 나중 탭이 앞선 방의 열쇠로
		// 간다. 노드 3개에서 재 보니 69.1% 가 엉뚱한 노드로 갔다.
		//
		// 경로에 실으면 요청마다 다르다. SockJS 가 기본 주소 뒤에 경로를 덧붙이므로
		// 쿼리스트링은 살아남지 못하지만, 기본 주소의 경로는 그대로 남는다.
		//
		// 옛 주소도 함께 남긴다. 배포 중에는 이미 붙어 있는 연결과 옛 화면이 있다.
		registry.addEndpoint("/ws/chat", "/ws/chat/{roomId}")
			.setAllowedOriginPatterns(allowedOrigins.toArray(new String[0]))
			.withSockJS()
			.setStreamBytesLimit(512 * 1024)
			.setHttpMessageCacheSize(1000)
			.setDisconnectDelay(30 * 1000)
			.setHeartbeatTime(25 * 1000);
	}

	/**
	 * 크기와 시간에 상한을 둔다. 갑자기 몰릴 때 서버가 먼저 죽지 않게 하기 위해서다.
	 */
	@Override
	public void configureWebSocketTransport(WebSocketTransportRegistration registry) {
		// 이미지와 파일은 따로 올리므로 메시지 자체는 클 이유가 없다
		registry.setMessageSizeLimit(128 * 1024)
			.setSendBufferSizeLimit(512 * 1024)
			.setSendTimeLimit(20 * 1000)
			.setTimeToFirstMessage(30 * 1000);
	}

	/**
	 * 클라이언트에서 서버로 들어오는 채널.
	 *
	 * <p>사람이 타이핑하는 속도라 빠르지 않다. 풀을 크게 잡을 이유가 없다.
	 */
	@Override
	public void configureClientInboundChannel(ChannelRegistration registration) {
		registration
			.interceptors(webSocketAuthInterceptor)
			// 평범한 풀 대신 연결 단위로 묶인 실행기를 쓴다. 풀에 그대로 던지면 한
			// 연결에서 연달아 보낸 프레임이 서로 앞질러, 나중에 보낸 것이 먼저 번호를
			// 받는다. 보낸 사람 화면에서도 자기 말의 앞뒤가 바뀐다.
			.executor(new SessionOrderedExecutor(
				new KeyOrderedExecutor(
					Math.max(2, Runtime.getRuntime().availableProcessors() * 2),
					"ws-inbound-")));
	}

	/**
	 * 서버에서 클라이언트로 나가는 채널.
	 *
	 * <p>한 건이 들어오면 두 사람에게 나간다. 여러 방에서 동시에 오가면 순간적으로
	 * 몰리므로 들어오는 쪽보다 넉넉하게 잡는다.
	 */
	@Override
	public void configureClientOutboundChannel(ChannelRegistration registration) {
		registration.taskExecutor()
			.corePoolSize(10)
			.maxPoolSize(50)
			.queueCapacity(1000)
			.keepAliveSeconds(60);
	}
}

File: backend/src/main/kotlin/com/joying/chat/config/WebSocketConfig.kt
package com.joying.chat.config

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import org.springframework.messaging.simp.config.ChannelRegistration
import org.springframework.messaging.simp.config.MessageBrokerRegistry
import org.springframework.web.socket.config.annotation.EnableWebSocketMessageBroker
import org.springframework.web.socket.config.annotation.StompEndpointRegistry
import org.springframework.web.socket.config.annotation.WebSocketMessageBrokerConfigurer
import org.springframework.web.socket.config.annotation.WebSocketTransportRegistration

/**
 * WebSocket STOMP 설정
 *
 * 실시간 채팅을 위한 WebSocket 설정
 * - STOMP 프로토콜 사용
 * - SockJS Fallback 지원
 * - Rate Limiting 및 연결 제한
 * - Backpressure 설정
 */
@Configuration
@EnableWebSocketMessageBroker
class WebSocketConfig(
    @Value("\${cors.allowed-origins}")
    private val allowedOrigins: List<String>,
    private val webSocketAuthInterceptor: WebSocketAuthInterceptor
) : WebSocketMessageBrokerConfigurer {

    /**
     * Message Broker 설정
     *
     * - Application Prefix: /app (클라이언트 → 서버)
     * - Redis Pub/Sub: 서버 간 메시지 전달
     *
     * SimpleBroker 제거 이유:
     * - SimpleBroker는 구독 정보를 서버 메모리에 저장
     * - 서버 확장 시 다른 서버의 구독자에게 메시지 전달 불가
     * - Redis Pub/Sub으로 직접 구독 관리하여 확장성 확보
     */
    override fun configureMessageBroker(registry: MessageBrokerRegistry) {
        // Application Destination Prefix
        // 클라이언트가 메시지를 보낼 때 사용
        // 예: SEND /app/chat/123/send
        registry.setApplicationDestinationPrefixes("/app")

        // SimpleBroker 제거 (Redis Pub/Sub으로 대체)
        // - 기존: registry.enableSimpleBroker("/topic", "/queue")
        // - 변경: ChatMessageListener에서 직접 WebSocket 전송

        // User Destination Prefix
        // 특정 사용자에게만 메시지 전송 시 사용
        // 예: /user/{username}/queue/messages
        registry.setUserDestinationPrefix("/user")
    }

    /**
     * STOMP Endpoint 등록
     *
     * - Endpoint: /ws/chat
     * - SockJS Fallback 지원 (WebSocket 미지원 브라우저 대응)
     * - CORS 허용
     */
    override fun registerStompEndpoints(registry: StompEndpointRegistry) {
        registry
            .addEndpoint("/ws/chat")  // WebSocket 연결 endpoint
            .setAllowedOriginPatterns(*allowedOrigins.toTypedArray())  // CORS 허용
            .withSockJS()  // SockJS Fallback 활성화
            .setStreamBytesLimit(512 * 1024)  // 스트리밍 전송 크기 제한: 512KB
            .setHttpMessageCacheSize(1000)  // HTTP 메시지 캐시 크기: 1000개
            .setDisconnectDelay(30 * 1000)  // 연결 끊김 지연 시간: 30초
            .setHeartbeatTime(25 * 1000)  // Heartbeat 주기: 25초
    }

    /**
     * WebSocket Transport 설정
     *
     * 메시지 크기 제한, 전송 버퍼 크기 제한, 타임아웃 설정
     * → 갑작스러운 부하 증가 시 서버 보호
     */
    override fun configureWebSocketTransport(registry: WebSocketTransportRegistration) {
        registry
            .setMessageSizeLimit(128 * 1024)  // 단일 메시지 최대 크기: 128KB (이미지/파일은 별도 업로드)
            .setSendBufferSizeLimit(512 * 1024)  // 전송 버퍼 크기: 512KB
            .setSendTimeLimit(20 * 1000)  // 전송 타임아웃: 20초
            .setTimeToFirstMessage(30 * 1000)  // 첫 메시지 대기 시간: 30초 (CONNECT 이후)
    }

    /**
     * Inbound Channel 설정 (클라이언트 → 서버)
     *
     * 사용자 입력 메시지를 받아서 처리
     *
     * 부하 특성:
     * - 사람 타이핑 속도: 초당 5-10개 메시지 (최대)
     * - MongoDB 저장 + Redis Pub/Sub 발행 (I/O 작업)
     * - 상대적으로 느리지만 CPU 집약적이지 않음
     *
     * 최적화 전략:
     * - 작은 Thread Pool (20개)로 충분
     * - 큐는 중간 크기 (200개) - 급격한 스파이크 대비
     * - keepAlive 짧게 (30초) - 유휴 시 빠른 해제
     */
    override fun configureClientInboundChannel(registration: ChannelRegistration) {
        registration
            // 쿠키 기반 인증 인터셉터 등록
            .interceptors(webSocketAuthInterceptor)
            .taskExecutor()
            .corePoolSize(5)  // 기본 5개 (유휴 시 최소 스레드)
            .maxPoolSize(20)  // 최대 20개 (사용자 입력은 느림)
            .queueCapacity(200)  // 대기 큐 200개 (갑작스러운 동시 전송 대비)
            .keepAliveSeconds(30)  // 30초 후 유휴 스레드 해제
    }

    /**
     * Outbound Channel 설정 (서버 → 클라이언트)
     *
     * Redis Pub/Sub에서 받은 메시지를 클라이언트에게 전송
     *
     * 부하 특성:
     * - 1:1 채팅: 1개 메시지 → 2명에게 전송 (본인 + 상대방)
     * - 네트워크 I/O 위주 (CPU는 낮음)
     * - 처리 속도 빠름 (5ms/건)
     * - 동시에 여러 채팅방에서 메시지 발생 시 순간 폭증
     *
     * 최적화 전략:
     * - Inbound 대비 2.5배 Thread Pool (50개) - 메시지 2건 + 버스트 대비
     * - 큰 큐 (1000개) - 동시 브로드캐스트 병목 방지
     * - keepAlive 길게 (60초) - 스레드 재사용 극대화
     */
    override fun configureClientOutboundChannel(registration: ChannelRegistration) {
        registration.taskExecutor()
            .corePoolSize(10)  // 기본 10개 (항상 대기)
            .maxPoolSize(50)  // 최대 50개 (Inbound 대비 2.5배, 1:1 채팅 + 버스트 대비)
            .queueCapacity(1000)  // 큐 1000개 (동시 브로드캐스트 병목 방지)
            .keepAliveSeconds(60)  // 60초 유지 (재사용 극대화)
    }
}